src/Servers/Kestrel/Transport.Abstractions/src/Internal/TransportConnection.FeatureCollection.cs

```diff
@@ -100,5 +105,68 @@ void IConnectionHeartbeatFeature.OnHeartbeat(System.Action<object> action, objec
         {
             OnHeartbeat(action, state);
         }
+
+        public void OnCompleted(Func<object, Task> callback, object state)
+        {
+            if (_onCompleted == null)
+            {
+                _onCompleted = new Stack<KeyValuePair<Func<object, Task>, object>>();
+            }
+            _onCompleted.Push(new KeyValuePair<Func<object, Task>, object>(callback, state));
+        }
+
+        public Task FireOnCompleted()
```

Comment:
I don't love FireOnCompleted() being public since it's not callled/needed by the transports that might reference Transport.Abstractions. I guess the same could be said for various property setters.

Out of caution, I think we should throw if this gets called multiple times. Or if someone calls OnCompleted after FireOnCompleted was already called.

Reply:
Another alternative is exposing the callbacks themselves via the TransportConnection and let the ConnectionDispatcher invoke them directly. This neatly solves the logging problem, but everything else about it is worse IMHO.

Reply:
`TrySetCompleted`? :)

Reply:
Or just `Complete`/`TryComplete`, since generally that's the naming pattern we've used in other places.